File: largestTreeRow.py
# Time Complexity : O(1)
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this :

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def largestValues(self, root: TreeNode) -> List[int]:
        if root is None:
            return []
        
        res = []
        q = deque()
        level = 0
        
        q.append(root)
        while len(q) != 0:
            size = len(q)
            
            for _ in range(size):
                curr = q.popleft()
                if level == len(res):
                    res.append(curr.val)
                else:
                    res[level] = max(res[level], curr.val)
                    
                if curr.left is not None:
                    q.append(curr.left)
                if curr.right is not None:
                    q.append(curr.right)
            level += 1
        return res

File: largest_treerow.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

// Time Complexity : O(N)
// Space Complexity : O(h) h is height of recursive stack
// Did this code successfully run on Leetcode : Yes

class Solution {
      public List<Integer> largestValues(TreeNode root) {
          List<Integer> result = new ArrayList<>();
          if(root == null)
              return result;
          dfs(root, result,  0);
          
          return result;
      }
      private void dfs(TreeNode root, List<Integer> res, int d){
          if(root == null)
              return;
          
            if(d == res.size()){
                res.add(root.val);
            }
            else{
                res.set(d, Math.max(res.get(d), root.val));
            }
          dfs(root.left, res, d+1);
          dfs(root.right, res, d+1);
      }
}


/*
// bfs solution
// Time Complexity : O(N)
// Space Complexity : O(N) 
// Did this code successfully run on Leetcode : Yes

class Solution {
      public List<Integer> largestValues(TreeNode root) {
          List<Integer> result = new ArrayList<>();
          if(root == null)
              return result;
          Queue<TreeNode> q = new LinkedList<>();
          q.add(root);
          int rindex = 0;
          while(!q.isEmpty()){
              int size = q.size();
              int max = Integer.MIN_VALUE;
              for(int i=0; i<size; i++){
                  TreeNode temp = q.poll();
                  max = Math.max(temp.val, max);
                  if(temp.left != null)
                      q.add(temp.left);
                  if(temp.right != null)
                      q.add(temp.right);
              }
              result.add(rindex, max);
              rindex++;
          }
          return result;
      }
}
*/